cpp/src/arrow/array/data.h

```diff
@@ -266,16 +266,19 @@ struct ARROW_EXPORT ArraySpan {
   int64_t offset = 0;
   BufferSpan buffers[3];
 
+  // 16 bytes of scratch space to enable this ArraySpan to be a view onto
+  // scalar values including binary scalars (where we need to create a buffer
+  // that looks like two 32-bit or 64-bit offsets)
+  uint8_t scratch_space[16];
```

Comment:
It might make sense to put this "scratch space" only in the Scalar values that require it (types that have offsets), then we don't have to carry around an extra 16 bytes with every ArraySpan

Reply:
Scalars that are a view onto a variable-size array (list, large list, binary, utf8, large binary, etc.) do not have an offsets buffer in their representation

Reply:
I'm making this aligned (by changing the type to be two uint64's)

Reply:
You can also use [`alignas`](https://en.cppreference.com/w/cpp/language/alignas)